doc/crypto/api/ops/ka.rst

```diff
@@ -114,6 +122,89 @@ Key agreement algorithms
 Standalone key agreement
 ------------------------
 
+.. function:: psa_key_agreement
+
+    .. summary::
+        Perform a key agreement and return the shared secret as a derivation key.
+
+    .. param:: psa_key_id_t private_key
+        Identifier of the private key to use.
+        It must permit the usage `PSA_KEY_USAGE_DERIVE`.
+    .. param:: const uint8_t * peer_key
+        Public key of the peer. The peer key must be in the same format that `psa_import_key()` accepts for the public key type corresponding to the type of ``private_key``. That is, this function performs the equivalent of :code:`psa_import_key(..., peer_key, peer_key_length)`, with key attributes indicating the public key type corresponding to the type of ``private_key``. For example, for ECC keys, this means that peer_key is interpreted as a point on the curve that the private key is on. The standard formats for public keys are documented in the documentation of `psa_export_public_key()`.
+    .. param:: size_t peer_key_length
+        Size of ``peer_key`` in bytes.
+    .. param:: psa_algorithm_t alg
+        The key agreement algorithm to compute: a value of type `psa_algorithm_t` such that :code:`PSA_ALG_IS_RAW_KEY_AGREEMENT(alg)` is true.
+    .. param:: const psa_key_attributes_t * attributes
+        The attributes for the new key.
+        This function uses the attributes as follows:
+
+        *   The key type must be one of `PSA_KEY_TYPE_DERIVE`, `PSA_KEY_TYPE_RAW_DATA`, `PSA_KEY_TYPE_HMAC`, or `PSA_KEY_TYPE_PASSWORD`.
+
+            Implementations must support the `PSA_KEY_TYPE_DERIVE` and `PSA_KEY_TYPE_RAW_DATA` key types. Support for output as `PSA_KEY_TYPE_HMAC` or `PSA_KEY_TYPE_PASSWORD` is :scterm:`implementation defined` .
+
+        *   The key size is always determined from the key agreement's shared secret. If the key size in ``attributes`` is nonzero, it must be equal to the size of the shared secret, in bits.
```

Comment:
Missing: saying explicitly that if the bit-size is specified as 0, it's the size of the shared secret.

Reply:
Happy to add that clarification.

Note that the wording here was derived from that used in `psa_import_key()` and `psa_copy_key()`:
* `psa_import_key()` says that the 'size is determined from the import data. If non-zero, ...' but does state in the full description the effect of a zero size in the attributes.
* `psa_copy_key()` says that the 'size can be zero. If non-zero, ...' with no further explanation anywhere.

An alternative would be to require that the size MUST be zero, as there is only one valid non-zero value, and it takes time and effort for the application to set it correctly, and for the implementation to validate it?

Reply:
If the application knows exactly what the expected attributes are, it should be allowed to pass them. In particular, it should be possible to call `psa_get_key_attributes` and reuse the structure to create a new key with the same attributes (only changing the key id if one of the keys is persistent).

Reply:
Ok, fair use case.